src/tomography.jl

```diff
@@ -0,0 +1,244 @@
+
+
+ρr = Variable(2^n, 2^n)
+ρi = Variable(2^n, 2^n)
+
+constraints = trace(ρr) == 1
+constraints += trace(ρi) == 0
+constraints += isposdef([ρr ρi; -ρi ρr])
+
+#We want to minimize the difference between predicted results and experimental results
+problem = minimize( vecnorm(invVarMat*(expResults - [real(predictorMat) imag(predictorMat)]*[vec(ρr); vec(ρi)]), 2)^2, constraints )
+solve!(problem, solver)
+
+@printf("Done\n")
+return (ρr.value - 1im*ρi.value)
+end
+
+
+function analyzeStateTomo2(data, nbrQubits, nbrPulses, nbrCalRepeats)
+
+    numMeas = length(data);
+    measOps = Dict();
+    tomoData = Array(Float64[]);
+    varData = Array(Float64[]);
+    if haskey(data, "data")
+        datatemp = Dict()
+        datatemp[1]=data
+        numMeas=1
+    end
+for ct = 1:numMeas
+    #Average over calibration repeats
+    calData = mean(reshape(real(datatemp[ct]["data"][end-nbrCalRepeats*(2^nbrQubits)+1:end]), nbrCalRepeats, 2^nbrQubits), 1);
+    #Pull out the calibrations as diagonal measurement operators
+    measOps[ct] = diagm(calData[:]);
+
+    #The data to invert
+
+        tomoData = [tomoData;  real(datatemp[ct]["data"][1:end-nbrCalRepeats*(2^nbrQubits)])];
+
+    #variance
+        varData = [varData; datatemp[ct]["realvar"][1:end-nbrCalRepeats*(2^nbrQubits)]];
+
+end
+#Map each experiment to the appropriate readout pulse
+    measOpMap = reshape(repmat(transpose(1:numMeas), nbrPulses^nbrQubits, 1), numMeas*(nbrPulses^nbrQubits), 1);
+
+    measPulseMap = repmat((1:nbrPulses^nbrQubits), numMeas, 1);
+#Use a helper to get the measurement unitaries.
+measPulseUs = tomo_gate_set(nbrQubits, nbrPulses);
+varMat = diagm(varData[:]);
+
+#Now call the inversion routines
+
+#First least squares
+    rhoLSQ = QST_LSQ(tomoData, varMat, measPulseMap, measOpMap, measPulseUs, measOps, nbrQubits);
+#plotting to be implemented    pauliSetPlot(rho2pauli(rhoLSQ), newplot);
+
+#Now constrained SDP
+    rhoSDP = QST_SDP(tomoData, varMat, measPulseMap, measOpMap, measPulseUs, measOps, nbrQubits);
+
+    return (rhoLSQ, rhoSDP)
+end
+
+function rho2pauli(rho)
+#converts a density matrix to a Pauli set
+
+nbrQubits = Int(log2(size(rho,1)));
+    kronMat = zeros(Int, 4^nbrQubits, nbrQubits);
+for qubitct = 1:nbrQubits
+    kronMat[:,qubitct] = reshape(repmat((0:3)', 4^(nbrQubits-qubitct), 4^(qubitct-1)), 4^nbrQubits, 1);
+end
+
+pauliVec = ones(4^nbrQubits,1);
+    for ii = 1:4^nbrQubits
+        pauliVec[ii] = real(trace(rho*pauli(kronMat[ii,:][:])));
+end
+
+    pauliStrs1Q = ["I","X","Y","Z"]
+    pauliStrs = cell(4^nbrQubits, 1);
+
+for paulict = 1:4^nbrQubits
+        pauliStrs[paulict] = "";
+    for qubitct = 1:nbrQubits
+        pauliStrs[paulict] = pauliStrs[paulict]*pauliStrs1Q[kronMat[paulict, qubitct]+1];
+    end
+    end
+    pauliVec =  pauliVec[sortperm(map(pauliHamming,pauliStrs)[:])]
+    pauliStrs = pauliStrs[sortperm(map(pauliHamming,pauliStrs)[:])] #reorder
+    return pauliVec, pauliStrs
+end
+
+function pauliHamming(str)
```

Comment:
This function also exists in Cliffords.jl and is just called `weight`, i.e. `weight(kron(X,X)) == 2`.